Syntax Highlighting Plugin

radian628 opened the pull request:
Adds a little plugin I made for syntax highlighting, containing the following features:
- Bracket pair colorization (with an option to also color the text), supporting any number of custom colors in any order.
- Increase the bracket thickness to make the colorization more visible.
- Highlights the nearest bracket pair to the caret, and the nearest bracket pair to the mouse pointer.

jared-hughes commented:
I love this! Styling happens instantly, and it always feels correct. I'm happy it works in all math input fields.

Many thoughts:
- Mouse cursor hover should not trigger the highlighting. I found it surprising when both the mouse hovered one bracket block, and the keyboard cursor hovered another.
  - Another option, if you want to keep mouse cursor hover: allow mouse cursor hover to highlight, but only if mouse movement was the last event input. Otherwise, allow only the mathquill cursor. But that's some tricky logic.
- Underlining should not be enabled by default. It felt like too much going on with the thickness and the color and the gray background and the underline.
- Turning off "Highlight Bracket Blocks" seems to be temporary. Try turning it off then clicking around. The underline and gray background comes back.
- The upper bound of "Thicken Brackets" should be lowered, maybe to less than 50% of the current value.
- If "Color in Text" is enabled, then `,` might want to be thickened a little for yellow
- Even with "Color in Text" disabled, it would be nice to highlight `,` (in all three brackets), `:` (in curly braces), and `...` (in square brackets).
  - Though it sounds like this is hard since there's not much semantic info on the mq. And `...` could be especially hard because they're separate nodes.
  - Be cautious of `,` outside of brackets, like from actions or from top-level `(1,2),(3,4)` being a list of points. I *think* these may as well be colored as well, but no preference.
- Maybe it's better to just get rid of yellow. There's a reason yellow is not part of the normal color cycle (bad contrast). Replace it with blue? Or just get rid of it.
  - Brown also can look good. 
  - A cycle of three colors would suffice. Three is the minimum, four works too.
  - ![image](https://github.com/DesModder/DesModder/assets/20214911/6e06bb63-6cf6-48fc-a369-6a153d1995fd)
- If you get rid of yellow, "thicken brackets" can be disabled by default.
- Underlines cause resizing (aka up-down jiggle). `margin-bottom: -1px` can fix this. (or, have a 1 pixel transparent border by default)

jared-hughes commented on localization/en.ftl:
```diff
@@ -246,3 +246,17 @@ custom-mathquill-config-opt-leftIntoSubscript-name = Left/Right into Subscripts
+syntax-highlighting-name = Syntax Highlighting
+syntax-highlighting-desc = Color in various parts of expressions to make them easier to reason about.
+syntax-highlighting-opt-bracketPairColorization-name = Bracket Pair Colorization
+syntax-highlighting-opt-bracketPairColorization-desc = Applies a set of alternating colors to brackets (e.g. ()[]{}||) to make matching bracket pairs easy to spot.
```
`{}` is ftl syntax, so "Error formatting key syntax-highlighting-opt-bracketPairColorization-desc".  I think you fix it with `{"{}"}`

radian628 replied:
fixed

jared-hughes commented on src/plugins/pillbox-menus/components/Menu.less:
```diff
@@ -110,6 +110,73 @@
+      width: 100%;
+    }
+
+    // for some reason these only work if separated
```
Weird. I would chalk it up to one of them erroring taking down the whole rule, but they're both `-webkit`, so neither should error.

radian628 replied:
Yeah tbh I don't remember exactly why it was broken but I think the actual reason was unrelated.

jared-hughes commented on src/plugins/pillbox-menus/components/Menu.tsx:
```diff
@@ -193,6 +198,95 @@ export default class Menu extends Component<{
+  plugin: SpecificPlugin,
+  settings: GenericSettings
+) {
+  // const clItem = item as ConfigItemColorList;
+
```
```suggestion
```

jared-hughes commented on src/plugins/pillbox-menus/components/Menu.tsx:
```diff
@@ -193,6 +198,95 @@ export default class Menu extends Component<{
+                  (settings[item.key] as string[]).map((e, i) => [e, i])
+                }
+                // TODO: find a better way of dealing with this
+                key={() => counter++}
```
This can probably be `key={([e,i]: [string, number]) => i}`.
If that doesn't update on change, then I suggest using instead `<For each={() => (settings[item.key] as string[]).map((e, i) => i)}>`, then Inside the For, replace `value={v}` with `value={() => settings[item.key][i]}`

radian628 replied:
As for your first suggestion, that doesn't seem to work correctly as far as I can tell--- all appends and deletions seem to happen to the end of the list instead of at the intended location.

The second one would probably work if not for the fact that in theory someone could put duplicate colors.

radian628 replied:
I might be able to get away with making some sort of a "hybrid key" made of both the index and the hex code. Ideally each element would have its own key that'd "stick around" with it as it moved through the collection, though that'd make things rather annoying to implement unless I decided to expose that functionality of "keyed arrays" directly to the plugins.

jared-hughes commented on src/plugins/syntax-highlighting/bracket-pair-colorization.ts:
```diff
@@ -0,0 +1,47 @@
+) {
+  if (!colors[0]) return [];
+
+  const colorMaker = `rgb(${colors[0]
+    .map((_, i) => {
+      return `calc(${colors
+        .map((col, colindex) => {
+          const channel = col[i];
+          return `${channel} * max(0, 10000 * cos(
+            2 * 3.14159265358979323 * (var(--test2) - ${colindex}) / ${colors.length}
+            ) - 9999)`;
+        })
+        .join(" + ")})`;
+    })
+    .join(", ")})`;
```
This took a bit for me to figure out. Might want to explain how the periodic cosine effects the color cycling. The scale to [-19999,1] (then clamping to non-negative) is used for what? Ensuring that anything slightly off-peak is zero?

radian628 replied:
That's correct. I'll add a comment clarifying, nice catch.

radian628 replied:
Done.

jared-hughes commented on src/plugins/syntax-highlighting/bracket-pair-colorization.ts:
```diff
@@ -0,0 +1,47 @@
+  return [
+    `
+    .dcg-mq-root-block {
+        --test1: 0;
+        --test2: 0;
```
Do these have more descriptive names than `test`?

radian628 replied:
Yeah, those are placeholders that I forgot to change lol

radian628 replied:
fixed

radian628 commented:
Thanks for the feedback! 

> Mouse cursor hover should not trigger the highlighting. I found it surprising when both the mouse hovered one bracket block, and the keyboard cursor hovered another. Another option, if you want to keep mouse cursor hover: allow mouse cursor hover to highlight, but only if mouse movement was the last event input. Otherwise, allow only the mathquill cursor. But that's some tricky logic.

Maybe I should make "mouse cursor hover highlighting" (with a shorter name lol--- maybe "hover highlighting"?) an optional setting, disabled by default. As for when to enable the hover highlighting, I like your idea of having it only when mouse movement was the last event (though like you implied it might be challenging to exhaustively cover all events), but alternatively it might be good to just make it mutually exclusive with the MQ cursor--- i.e. the hover highlight is always enabled while hovering unless the MQ cursor highlight is enabled.

> Underlining should not be enabled by default. It felt like too much going on with the thickness and the color and the gray background and the underline.

Yep, agreed. I'll disable that by default and gate it under a setting.

> Turning off "Highlight Bracket Blocks" seems to be temporary. Try turning it off then clicking around. The underline and gray background comes back.

Weird, I'll test that.

> The upper bound of "Thicken Brackets" should be lowered, maybe to less than 50% of the current value.

Agreed.

> Even with "Color in Text" disabled, it would be nice to highlight , (in all three brackets), : (in curly braces), and ... (in square brackets). Though it sounds like this is hard since there's not much semantic info on the mq. And ... could be especially hard because they're separate nodes.

Oohhh, I like this, cuz they're kind of part of the enclosing context. I'll definitely add this when I get the chance. From what I've learned with Multiline I think it should be partially doable by analyzing the HTML, though RIP CSS-only bracket pair colorization :(.

> Be cautious of , outside of brackets, like from actions or from top-level (1,2),(3,4) being a list of points. I think these may as well be colored as well, but no preference.

Oh yeah, this does present an issue that might make figuring out what commas to highlight harder. For instance, comma-separated actions inside of piecewises versus commas for the piecewise itself.

> If "Color in Text" is enabled, then , might want to be thickened a little for yellow
> Maybe it's better to just get rid of yellow. There's a reason yellow is not part of the normal color cycle (bad contrast). Replace it with blue? Or just get rid of it.
> Brown also can look good.
> If you get rid of yellow, "thicken brackets" can be disabled by default.

Yep, agreed. I'll change yellow to something else.

> A cycle of three colors would suffice. Three is the minimum, four works too.

For a normal code editor I'd 100% agree. However, since Desmos relies so much more on shape than on color to convey meaning, we have a lot more of the color wheel to use for bracket pair colorization and disambiguating very deeply-nested expressions. That being said, 3 or 4 colors is probably sufficient either way.

> Underlines cause resizing (aka up-down jiggle). margin-bottom: -1px can fix this. (or, have a 1 pixel transparent border by default)

Sounds good, I'll implement it.

jared-hughes commented on src/plugins/syntax-highlighting/index.less:
```diff
@@ -0,0 +1,9 @@
+*[data-contains-caret="true"] {
```
The `*` here (and in the next one) is unnecessary.

radian628 replied:
fixed

radian628 commented:
Ooohh, another thought I just had: If I can figure out how to parse them, bracket pair colorization could also apply to integrals.

jared-hughes commented on src/plugins/syntax-highlighting/index.ts:
```diff
@@ -0,0 +1,96 @@
+  ];
+}
+
+export function toggleRules(rules: string[], toggle: boolean) {
+  return toggle ? rules : [];
+}
+
```
This function seems unused.

radian628 replied:
ah, the good 'ol speculative abstraction

radian628 replied:
fixed

jared-hughes commented on src/plugins/syntax-highlighting/index.ts:
```diff
@@ -0,0 +1,96 @@
+import { Config, configList } from "./config";
+import "./index.less";
+
+// assumes valid input;
+export function hex2rgb(hex: string): [number, number, number] {
+  return [
+    parseInt(hex.slice(1, 3), 16),
+    parseInt(hex.slice(3, 5), 16),
+    parseInt(hex.slice(5, 7), 16),
+  ];
+}
```
You may be interested in `parseCSSHex` from GLesmos `colorParsing` (which could probably be moved to a utils somewhere). Such a refactor is probably only helpful if invalid input is possible.

radian628 replied:
Invalid input should be impossible here so I think I'm fine on that front.

jared-hughes commented on src/plugins/syntax-highlighting/index.ts:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    for (const rule of bpcCss) {
+      this.styles.sheet?.insertRule(rule);
```
This API is beautiful. I've just innerHTML'd my way into styles in the past.

jared-hughes commented:
- On Firefox, you can click one swatch option, then the next one, then the next one, and it pops up a bunch of menus
![image](https://github.com/DesModder/DesModder/assets/20214911/d7a332a5-3a64-4e04-b132-7655bff31dc4)

- > Turning off "Highlight Bracket Blocks" seems to be temporary. Try turning it off then clicking around. The underline and gray background comes back.

   Relatedly, turning off the plugin doesn't get rid of the gray background.

jared-hughes commented on src/plugins/syntax-highlighting/index.ts:
```diff
@@ -0,0 +1,96 @@
+    document.head.appendChild(this.styles);
+    this.afterConfigChange();
+
+    setInterval(() => {
```
I begrudge this `setInterval` since it handles all current and future mathquill fields, unlike hooking into dispatched events.

radian628 replied:
Hold on, what do you mean by "handles all current and future mathquill fields"? I realize now that in theory I could just replace it with a listener that listens for _all_ events without filtering any of them.

jared-hughes replied:
That wouldn't work for the DesModder math fields.

radian628 replied:
Ahh, true, I see.

jared-hughes commented:
> Oh yeah, this does present an issue that might make figuring out what commas to highlight harder. For instance, comma-separated actions inside of piecewises versus commas for the piecewise itself.

Commas inside a piecewise are always separating piecewise branches. Joining actions with commas requires parentheses.

> Since Desmos relies so much more on shape than on color to convey meaning, we have a lot more of the color wheel to use for bracket pair colorization

Good point, though we want to leave open the option of coloring identifiers, numbers, function names, etc.

radian628 commented on src/plugins/pillbox-menus/components/Menu.tsx:
```diff
@@ -1,13 +1,16 @@
 import PillboxMenus from "..";
+import "./Menu.css";
```
fixed

radian628 commented on src/plugins/pillbox-menus/components/Menu.css:
fixed

radian628 commented on src/plugins/index.ts:
```diff
@@ -51,11 +52,16 @@ export interface ConfigItemNumber extends ConfigItemGeneric {
   | ConfigItemString
-  | ConfigItemNumber;
+  | ConfigItemNumber
+  | ConfigItemColorList;
```
Thanks!

radian628 commented:
> On Firefox, you can click one swatch option, then the next one, then the next one, and it pops up a bunch of menus

I wonder if this is a problem with Firefox's color picker on Linux rather than a problem with the plugin. I can't seem to replicate this on Firefox on Windows 11. 

That being said, Firefox on Windows 11 has another issue lol: 
![image](https://github.com/DesModder/DesModder/assets/70352880/6ef4a8b5-e75c-44f6-bc81-c81725658e67)

jared-hughes commented:
![image](https://github.com/DesModder/DesModder/assets/20214911/6c821f3b-8c9f-4eb0-83dd-d4e21f110b48)
Mine looks right on linux. Guessing it supported the webkit pseudo-element. Firefox 116.0.1

radian628 commented:
Here's the fixes I've implemented so far:
- Removed yellow from the color rotation.
- Changed the "thicken brackets" default to 0.
- Added the negative margin to prevent the underline from shifting layout.
- Gated hover highlights under a separate setting
- Highlighted ranges now disable when the settings or the plugin is disabled.

radian628 commented:
Stuff I still need to work out:
- Fixing the CSS on Windows 11 Firefox (seems like there's [this](https://developer.mozilla.org/en-US/docs/Web/CSS/::-moz-color-swatch) so I'll probably just use that)
- Figuring out how to handle when to have the hover highlight enabled. Having it disable after a click, scroll, or keyboard event would be a nice start, but like you implied earlier, I'm not sure if that's exhaustive enough.
- Removing the underline by default

radian628 commented:
![image](https://github.com/DesModder/DesModder/assets/70352880/3a8ff5b1-a58a-45c2-8441-b89aadb2af9e)

Wait a second, it's working in FF on Windows 11 now? I didn't even touch anything; why is it working!?

radian628 commented:
Fixes:
- Firefox seems to work now with the color swatches for some reason.
- Underlines on highlighted ranges are gated behind a setting and are disabled by default
- Mouseover range highlighting disappears when a key is pressed, the mouse is clicked, or the user scrolls.

jared-hughes commented:
I suspect you closed Firefox, and it updated by the time you next opened it.

radian628 commented:
Huh weird, I don't remember closing it between it not working and it working. That would be kind of funny though--- a browser update fixing my bug lol.

radian628 commented:
Should we hold off on merging this until color themes is done so we can move the bracket pair colorization colors to that instead? It'll probably be really easy to port it over so imo it's fine to push earlier now.

jared-hughes commented:
Proposal: remove the color choices (just use a fixed, hardcoded list), then we can merge. This way, we won't have to deal with migrating config later on.

jared-hughes commented:
I'll try to fix the one unresolved comment (about keying a <For>), then it should be good to merge.
Looking ahead, we might not want to merge the color options into color themes, just because it'll be a lot of (too many) color choices in one place. But either way it goes, as you say, it won't be too hard to port it later.

jared-hughes reviewed (approved):
I pushed some changes. Make sure to look over those. PR looks good to me!

- Commas are now highlighted unconditionally. There was no special handling needed for piecewises. Just for top-level lists like `(0,0),(1,2)`.
- Changed default colors. My main gripe was red since it felt like an error. So I replaced it with blue. Also green and purple match the colors in `Calc.colors`

One thing you might want to change:
- "Color in Text" might need a rename. At first glance, I thought it would refer to notes. Maybe "Colorize enclosed math", though I suspect you can come up with something a bit more elegant.

jared-hughes commented:
For the keying thing, it might be helpful to use the IndexFor component you created. Again doesn't allow duplicated keys, so can be tricky.

jared-hughes commented:
I like the new colors. Are you able to fix the conflicts? Then this could be merged to main.

radian628 commented:
Yeah! It' shouldn't be that bad I hope lol. I'll update you when the conflicts are fixed.

radian628 commented:
Should be up-to-date now.
